Zend/zend_API.c

```diff
@@ -2284,12 +2300,20 @@ ZEND_API int zend_register_functions(zend_class_entry *scope, const zend_functio
 			if (ctor->common.fn_flags & ZEND_ACC_STATIC) {
 				zend_error(error_type, "Constructor %s::%s() cannot be static", ZSTR_VAL(scope->name), ZSTR_VAL(ctor->common.function_name));
 			}
+
+			if (ctor->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+				zend_error_noreturn(E_CORE_ERROR, "Constructor %s::%s() cannot declare a return type", ZSTR_VAL(scope->name), ZSTR_VAL(ctor->common.function_name));
```

Comment:
Might this refactoring change bevahior? e.g. if we have both ctor and dtor which are both static, and ctor declares a return type, the original code would do, in order:
(a) `zend_error(error_type, "Constructor %s::%s() cannot be static", ...);`
(b) `zend_error(error_type, "Destructor %s::%s() cannot be static", ...);`
(c) `zend_error_noreturn(E_CORE_ERROR, "Constructor %s::%s() cannot declare a return type", ...);`
whereas now it would do (c) before (b), and assuming (c) aborts the execution (while (a) and (b) might not, esp. depending on `error_type`), (b) would not be emitted anymore... OTOH the other file `Zend/zend_compile.c` already does
(1) `zend_error_noreturn(E_COMPILE_ERROR, "Constructor %s::%s() cannot be static", ...);`
(2) `zend_error_noreturn(E_COMPILE_ERROR, "Constructor %s::%s() cannot declare a return type", ...);`
(3) `zend_error_noreturn(E_COMPILE_ERROR, "Destructor %s::%s() cannot be static", ...);`
which I guess shorts stop at (1), so I don't know...

Reply:
The behavior hasn't changed using the following test-case:

```php
<?php

class Foo {
    public static function __construct(): void {}
    public static function __destruct(): void {}
}
```

Example: https://3v4l.org/oi30v

_Assuming that with behavior you mean the final messages for userland._

Reply:
I guess that "uses" the other file `Zend/zend_compile.c`, see e.g. https://3v4l.org/aEuFY where you read "Clone method Foo::__clone() cannot be static" whereas with `Zend/zend_API.c` it would be just "Foo::__clone() cannot be static" (see also https://github.com/php/php-src/pull/3835#issuecomment-614499620)

Reply:
Maybe the checks on `zend_API` are related to PHP's extensions. If an extension declares a `__clone` method static, the check on `zend_API` will be used, while the `zend_compile` is used to compile everything else.

@nikic @cmb69 Mind to confirm this, please?

Reply:
Yes (internal vs. userland classes)

Reply:
Thanks. If this won't affect userland, feel free to resolve this digression 😄

Reply:
I'll align the messages, removing `Clone method...` and adopt `Method Foo::__clone`.